subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/BuildLogicLockingIntegrationTest.groovy

```diff
@@ -119,6 +124,110 @@ class BuildSourceBuilderIntegrationTest extends AbstractIntegrationSpec {
         runBlockingHandle?.abort()
     }
 
+    def "can simultaneously run gradle on projects with included build logic builds"() {
```

Comment:
💭  The test didn't really fail without the locking, so I am not sure what other things I'd need to introduce to make it fail.

Reply:
The data race window is probably very small (and I guess many overlapping executions would still be benign as long as the last writer gets to write a full valid `jar`) so many runs might be required before a problem can be observed (if ever with such small files).

Reply:
And it would make sense to also test some invariants like _`:pluginBuild:compileJava` must always be up-to-date in one of the builds_ when running builds in parallel.